### DISCUSSION TITLE
Places: How are common names determined? 🗺

### DISCUSSION THREAD

**Discussion by meonkeys** [photoprism/photoprism#1724] [Q&A] [answered]
How does photoprism match photos to place nicknames without reaching out to any online services? Is there an included database mapping lat/long to location names?

**Comment by lastzero** [comment, marked as answer]
See https://photoprism.app/features:
> Our privacy-preserving backend services provide worldwide location information to enrich your pictures with details such as state, city, and category. Future releases will be able to use public event data to automatically create albums of popular music festivals or sporting events.

Simply put, it's too much data to distribute... about 300 GB. Many of our users' servers also wouldn't have enough memory. There are no long-term logs and the backend doesn't know which pictures belong to what location, who you are, or even why the data is requested. Every smartphone sends more data home whenever you use a weather app.

Your local instance caches everything so that the same data isn't requested twice. Note PhotoPrism uses S2 cell IDs instead of precise lat / long floating point numbers, so it's never the exact location. As privacy friendly, cheap, and fast as it can get in practice.

See also https://twitter.com/photoprism_app/status/1458903521173123085

**Reply by meonkeys** [reply]
Thank you! And I saw `PHOTOPRISM_DISABLE_PLACES` is available as a [config option](https://docs.photoprism.app/getting-started/config-options/) to prevent PhotoPrism from reaching out to `places.photoprism.app`, should someone want even more privacy by disabling reverse geocoding altogether. That's handy.

Could you say a bit more about `places.photoprism.app`? What is running there?

**Reply by lastzero** [reply]
See documentation: https://docs.photoprism.org/developer-guide/metadata/geocoding/

Technology and tradeoffs have also been discussed in the comments:
- https://github.com/photoprism/photoprism/issues/1740
- https://github.com/photoprism/photoprism/issues/1776
- https://github.com/photoprism/photoprism/issues/1767

**Reply by lastzero** [reply]
Our backend now has more than 1 TB of geodata, up from 300 GB a month ago. And that's without backups and redundancy.

**Reply by meonkeys** [reply]
That is very helpful, thank you! That service/API does sound very privacy friendly and I certainly appreciate not having to self-host 1 TB of geodata.

Sorry to be a bother, just so I understand clearly: what is the software running at `places.photoprism.app`? Is it https://nominatim.org ? Or something else? I read what you shared but I couldn't figure out the answer to this. I apologize if I missed it somewhere.

**Reply by lastzero** [reply]
It's a custom S2 cell ID based Web service that has access to self-hosted Photon and Nominatim instances, with a fallback to external services in case of errors. I think it's discussed in those issues.

**Reply by meonkeys** [reply]
Gotcha. Is it possible to self-host the custom reverse geocoding web service? I'm guessing it's not, I'm just curious.

**Reply by lastzero** [reply]
Rest assured that we have absolutely no time to sell, store, or analyze any data... the last thing we need is extra work!  😂

As the GitHub commit history proves, we're busy implementing features and answering questions around the clock.... like right now and in the [chat at 2am](https://gitter.im/browseyourlife/community) this morning. Was the large one-time donation from you? 👀

**Reply by meonkeys** [reply]
Is it possible for me to self-host the custom reverse geocoding web service? The amount of storage space required is not an issue.

Let me also be very clear that I have no privacy concerns with PhotoPrism... you certainly sound trustworthy and I can see all the source that talks with the service. And if I don't want to use reverse geocoding you've made it very easy to disable.

Yes, I donated. You earned it! :+1:

I'm sorry to waste your time with this, I just want to make sure I really understand this backend service.

**Reply by lastzero** [reply]
> Gotcha. Is it possible to self-host the custom reverse geocoding web service? I'm guessing it's not, I'm just curious.

Yes, if you have at least $300 to spend on servers each month. With DigitalOcean or AWS, it's probably closer to $1000 per month.  
Also, we would have to find time to set it up for you and provide detailed documentation, which probably no one wants to pay for.

However:

- We have absolutely no time to maintain another open source project or write more documentation.
- Giving away enterprise-grade technology to businesses for free seems not the right thing to do when we're still losing money after 4 years of no personal life.
- There are limits to what is possible, healthy, and sustainable. We could use those $300+ per month well before anyone spends them on private reverse lookup servers. Seriously.

**Reply by lastzero** [reply]
Just saw your last reply.... thanks a lot! You're obviously welcome to join the sponsor chat with this 💎

**Reply by lastzero** [reply]
As for the hardware specifications, we are not simply talking about 1 TB of disk space. Our servers have 4 TB NVMe RAIDs, 128 GB RAM, and 32 threads. Importing OSM data could otherwise take weeks. AWS seems to charge $5,616.66 per month for this, if I'm not mistaken. Edit: Smaller servers of course will do for a personal instance.

**Reply by meonkeys** [reply]
All good. I hear you. It is not free to host the service and it is too much work to open source it. Makes sense, I agree, and I'm glad this choice helps y'all focus on the amazing core product. I don't expect you to work for free or provide optional features for free. Open sourcing the reverse geocoding service is not a huge deal to me because (a) I don't need nor want it and (b) the rest of PhotoPrism works just fine without it. I'll try to be your canary in the coal mine to make sure (b) is always true (as long as me doing so is helpful). Well, hopefully some unit/integration/acceptance tests do the heavy lifting here. :-)

I was asking because I'm interested in [FOSS](https://en.wikipedia.org/wiki/Alternative_terms_for_free_software) academically and ideologically, so I pay close attention to stuff like this. To me it doesn't necessarily matter if it is practical, I just like to know what is possible. An extreme example is Signal, which is almost entirely FOSS but is [challenging to fork in practice](https://github.com/LibreSignal/LibreSignal) (and this is by design). We're obviously not facing their same issues with PhotoPrism.

Thanks again for your helpful info, and for making an excellent product.

**Reply by lastzero** [reply]
In addition, the data needs frequent improvements. For example when users report bad state names as recently. 

It's much easier to maintain on a central instance than decentralized. If someone pays good money to install a private instance, why not? Doesn't make sense for the average user and that's not only because of hosting costs.
